src/test/java/com/google/devtools/build/lib/util/RegexUtilTest.java

```diff
@@ -0,0 +1,56 @@
+package com.google.devtools.build.lib.util;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.testing.junit.testparameterinjector.TestParameter;
+import com.google.testing.junit.testparameterinjector.TestParameterInjector;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/** Tests for {@link RegexUtil}. */
+@RunWith(TestParameterInjector.class)
+public class RegexUtilTest {
+
+  @Test
+  public void optimizedMatchingPredicate(
+      @TestParameter({
+            "",
+            "a",
+            "foo",
+            "foofoo",
+            "/coverage.dat",
+            "/coverage.data",
+            "/coverage1dat",
+            "/coverage1data",
+            "foo/coverage.dat",
+            "foo/coverage.data",
+            "foo/coverage1dat",
+            "foo/coverage1data",
+            "foo/test/a/coverage.dat",
+            "foo/test/.*/coverage.dat",
+            "]]\n",
+          })
+          String haystack,
+      @TestParameter({
+            ".*",
+            ".*?foo",
+            ".*+foo",
+            "^foo$",
+            ".*/coverage.dat",
+            ".*/coverage\\.dat",
+            ".*/test/.*/coverage\\.dat",
+            "$|",
+            "^",
+            ".]",
+            ".*]",
+            ".*^?^\\Q",
```

Comment:
Some more ideas for test cases:

- A regex with an alternation matching two different test strings
- A regex with at least one metacharacter following a `\Q`, with that character matching one of the test strings literally
- Test strings containing metacharacters not already covered above

Fuzzing is a great way to obtain additional confidence that the implementation is correct, but since it won't be triggered by CI, I'd still want some more test coverage.